hal/src/common/dmac/transfer.rs

```diff
@@ -0,0 +1,857 @@
+//! completed. When it returns, it will release the source and destination
+//! buffers, as well as the DMA channel and the payload.
+//!
+//! # Interrupting (stopping) a transfer
+//!
+//! A transfer can be stopped (regardless of whether it has completed or not) by
+//! calling [`stop`](DmaTransfer::stop). This is _not_ a blocking method,
+//! meaning it will stop the transfer and immediately return. When it returns,
+//! it will release the source and destination buffers, as well as the DMA
+//! channel and the payload.
+//!
+//! # Trigger sources
+//!
+//! Most peripherals can issue triggers to a DMA channel. A software trigger is
+//! also available (see [`trigger`](DmaTransfer::software_trigger)). See
+//! ATSAMD21 datasheet, table 19-8 for all available trigger sources.
+//!
+//! # Trigger actions
+//!
+//! Three trigger actions are available:
+//!
+//! * BLOCK: One trigger required for each block transfer. In the context of
+//!   this driver,
+//! one DmaTransfer is equivalent to one Block transfer.
+//!
+//! * BEAT: One trigger required for each beat transfer. In the context of this
+//!   driver, the beat
+//! size will depend on the type of buffer used (8, 16 or 32 bits).
+//!
+//! * TRANSACTION: One trigger required for a full DMA transaction. this is
+//!   useful for circular
+//! transfers in the context of this driver. One trigger will set off the
+//! transaction, that will now run uninterrupted until it is stopped.
+
+use core::{marker::PhantomData, sync::atomic};
+
+use super::{
+    channel::{Busy, Channel, Ready},
+    dma_controller::{DmaController, TriggerAction, TriggerSource},
+    DmacDescriptor, DESCRIPTOR_SECTION,
+};
+use crate::typelevel::Sealed;
+use core::mem;
+
+// TODO change source and dest types to Pin (see https://docs.rust-embedded.org/embedonomicon/dma.html#immovable-buffers)
+/// DMA transfer, owning the resources until the transfer is done and
+/// [`wait`](DmaTransfer::wait) is called.
+pub struct DmaTransfer<B, Pld, Src, Dst, ChanStatus, const ID: u8>
+where
+    B: 'static + DmaBeat,
+    Src: DmaBuffer<B>,
+    Dst: DmaBuffer<B>,
+{
+    xfer: UnsafeTransfer<B, Pld, Src, Dst, ChanStatus, ID>,
+}
+
+/// Unsafe DMA transfer, owning the resources until the transfer is done and
+/// [`wait`](DmaTransfer::wait) is called. The user *must* ensure the
+/// transfer won't be dropped until it `wait`ed upon or `stop`ped
+/// before the transfer is dropped.
+pub struct UnsafeTransfer<B, Pld, Src, Dst, ChanStatus, const ID: u8>
+where
+    B: DmaBeat,
+    Src: DmaBuffer<B>,
+    Dst: DmaBuffer<B>,
+{
+    config: TransferConfiguration<B, Src, Dst>,
+    chan: Channel<ChanStatus, ID>,
+    payload: Pld,
+}
+
+/// Transfer configuration abstraction. This is also used
+/// for `UnsafeTransfer`s so no `'static` bound.
+struct TransferConfiguration<B, Src, Dst>
+where
+    B: DmaBeat,
+    Src: DmaBuffer<B>,
+    Dst: DmaBuffer<B>,
+{
+    source: Src,
+    dest: Dst,
```

Comment:
Should `source` and `dest` be changed to `Pin<DmaBuffer<B>>`? [This](https://docs.rust-embedded.org/embedonomicon/dma.html#immovable-buffers) uses `Pin`ning but [this](https://blog.japaric.io/safe-dma/) doesn't. Can we just assume that a `DmaTransfer where B: 'static` is safe?